File: datastructure/doubleylinkedlist.js


class Node{
    constructor(value){
this.value=value
this.next=null
this.prev=null
    }
}



class doubleylinkedlist{
    constructor(){
        this.head=null
        this.tail=null
        this.size=0

    }

    isempty(){
        return this.size===0
    }
    getsize(){
        return this.size
    }
    prepend(value){
        let node=new Node(value)
        if(this.isempty()){
            this.head=node
            this.tail=node
        }else{
            node.next=this.head
            this.head.prev=node
            this.head=node
        }
        this.size++
    }
    removefromstart(){
        let value=0
        if(this.isempty()){
            return
        }
        
        if(this.size===1){
            value=this.head.value
            this.head=null
            this.tail=null
           
        }else{
            value=this.head.value
            let curr=this.head
            this.head=this.head.next
            curr.next=null
            this.head.prev=null  

        }
        this.size--
        return value
    }
    insert(index,value){
        if(index<0 || index>this.size){
            return 
        }
        if(index===0){
            this.prepend(value)
        }else{
            let node=new Node(value)
            let curr=this.head
            for(let i=0;i<index-1;i++){
                  curr=curr.next
            }
            let temp=curr.next
            curr.next=node
            node.prev=curr
            node.next=temp
            temp.prev=node
            this.size++
        }
    }
    removefromindex(index){
        if(index<0 || index>=this.size){
            return
        }
        if(index===0){
            this.removefromstart()
        }
        if(index===this.size-1){
            this.removefromend()
        } else{
            let curr=this.head
            for(let i=0;i<index-1;i++){
     curr=curr.next
            }
            let temp=curr.next
            let removed=curr.next
            temp=temp.next
            curr.next=temp
            temp.prev=curr
           removed.next=null
           removed.prev=null
           this.size--
        }
    }
    get(index){
      if(index<0 ||index>this.size-1){
        return console.log("index doesn't exist")
      }
      if(index===0){
        console.log(this.head.value)
        return this.head.value
      }else{
        let trav=this.head
        for(let i=0;i<index-1;i++){
            trav=trav.next
        }
              trav=trav.next
              console.log(trav.value)
              return trav.value
      }
      

    }
    sort(){
        let curr=this.head
        while(curr){ 
            let ahead=curr.next
            while(ahead){
                if(curr.value>ahead.value){
                    let temp=ahead.value
                    ahead.value=curr.value
                    curr.value=temp  
                }
                ahead=ahead.next
            }
            curr=curr.next
        }
    }

    append(value){
        let node=new Node(value)
        if(this.isempty()){
            
            this.head=node
            this.tail=node
        }else{
            node.prev=this.tail
            this.tail.next=node
            this.tail=node
        }
   this.size++
    }
    removefromend(){
        if(this.isempty()){
            return
        }
        let value=0
        if(this.getsize()===1){
            value=this.head.value
            this.head=null
            this.tail=null
        }else{
            value=this.tail.value
            this.tail=this.tail.prev
            this.tail.next=null
        }
        this.size--
        return value
    }
    print(){
        let curr=this.head
        let temp=""
        while(curr){
            temp+=`${curr.value} `
            curr=curr.next
        }
        console.log(temp)

    }
    printreverse(){
        let curr=this.tail
        let temp=""
        while(curr){
            temp+=`${curr.value} `
            curr=curr.prev
        }
        console.log(temp)
    }
}


module.exports=doubleylinkedlist

File: java-datastructure/doubleylinkedlist.java
public class doubleylinkedlist {
    private ListNode head;
    private ListNode tail;
    private int size;

    public static class ListNode {
        public int data;
        public ListNode next;
        public ListNode prev;


        public ListNode(int value){
            this.data=value;
            this.next=null;
            this.prev=null;
        }

    }

     public boolean isEmpty(){
        return this.size==0;
      }
     public int getSize(){
        return this.size;
     }

     public void append(int value){
        ListNode node=new ListNode(value);
        if(isEmpty()){
            this.head=node;
            this.tail=node;
        }else{
            node.prev=this.tail;
            this.tail.next=node;
            this.tail=node;
            this.tail.next=null;
        }
        this.size++;
     }
     public int removefromlast(){
        if(isEmpty()) return 404;
        ListNode curr=this.tail;
        this.tail=curr.prev;
        this.tail.next=null;
        curr.prev=null;
        this.size--;
        return curr.data;
     }

      public int removefromindex(int index){
        if(index<0 || index>=this.size){
            return 404;
        }
        if(index==0){
         return removefromfront();
        }
        if(index==this.size-1){
            return removefromlast();
        }
        if(index<=this.size/2){
            ListNode curr=head;
        ListNode remove=curr.next;
        for(int i=1;i<index;i++){
            curr=curr.next;
            remove=remove.next;
        }
        ListNode another=remove.next;
        curr.next=another;
        another.prev=curr;
        remove.next=null;
        remove.prev=null;
        this.size--;
        return remove.data;   
        }else{
            ListNode curr=tail;
            ListNode remove=curr.prev;
            for(int i=1;i<this.size-1-index;i++){
                curr=curr.prev;
                remove=remove.prev;
            }
            ListNode another=remove.prev;
              System.out.println("he0lo");
            curr.prev=another;
            another.next=curr;
            remove.next=null;
            remove.prev=null;
            this.size--;
           return remove.data; 
        } 
    }

     public String search(int value){
        if(isEmpty()) return "linkedlist is empty";
        if(this.head.data==value) return value+" is present on index 0";
        ListNode search=head;
        for(int i=0;i<this.size;i++){
            if(search.data==value) return value+" is present on index "+i;
            search=search.next;
        }
        return value+" is not present in linkedlist";
    }


    public int removefromfront(){
        if(isEmpty()) return 404;
        ListNode curr=this.head;
        this.head=curr.next;
        this.head.prev=null;
        curr.next=null;
        this.size--;
        return curr.data;
    }

     public void piepend(int value){
       ListNode node=new ListNode(value);
       if(isEmpty()){
        this.head=node;
        this.tail=node;
       }else{
           node.next=this.head;
           this.head.prev=node;
           this.head=node;
           this.head.prev=null;
       }
       this.size++;
     }

     public void print(){
        ListNode curr=head;
        String result="";
        while(curr!=null){
            result+=curr.data+" ";
            curr=curr.next;
        }
        System.out.println(result);
     }

     public static void main(String[] args) {
        doubleylinkedlist list=new doubleylinkedlist();
        list.append(1);
        list.append(2);
        list.append(3);
        list.append(4);
        list.append(5);
        list.append(6);
        list.append(7);
        list.append(8);
        list.append(9);
        list.append(10);
        System.out.println(list.getSize());
        list.removefromindex(4);
        list.removefromindex(7);
        list.print();
     }
          
    
}
